python/cudf/cudf/core/column/column.py

```diff
@@ -120,6 +120,34 @@ def data_array_view(self) -> "cuda.devicearray.DeviceNDArray":
         """
         return cuda.as_cuda_array(self.data).view(self.dtype)
 
+    @property
+    def _data_array_view(self) -> "cuda.devicearray.DeviceNDArray":
+        """
+        Internal implementation for viewing the data as a device array object
+        without triggering a deep-copy.
```

Comment:
> without triggering a deep-copy.

When would `.data_array_view` trigger a deep-copy? As a side note, `._mask.array_view` can still trigger a deep copy, as `col.mask` *must* copy data if the column's `.offset` attribute is not 0.

I think having 4 different properties `._data_array_view`, `.data_array_view`, `._mask_array_view` and `.mask_array_view` is confusing. When should we use which?

Reply:
Ditto below re: `.nullmask` and `._nullmask`

Reply:
> When would `.data_array_view` trigger a deep-copy?

`cuda.as_cuda_array` will trigger a call to `__cuda_array_interface__` which will trigger a deep-copy.


> I think having 4 different properties `._data_array_view`, `.data_array_view`, `._mask_array_view` and `.mask_array_view` is confusing. When should we use which?

It is confusing, but I felt the need to add these to avoid deep-copies for:
1. device to host conversions for example: `.values_host` which inturn calls `.data_array_view.copy_to_host()` 
2. Read only operations like:
```python
    a_col = column.as_column(a)
    a_array = cupy.asarray(a_col.data_array_view)

    b_col = column.as_column(b)
    b_array = cupy.asarray(b_col.data_array_view)

    result = cupy.isclose(
        a=a_array, b=b_array, rtol=rtol, atol=atol, equal_nan=equal_nan
    )
```

Reply:
I don't have a solution here, just chiming in to say that I also find this confusing and potentially concerning for future maintenance.

Reply:
Maybe we should be following the same pattern as `Buffer.ptr` and having a `mutable*` versions of these properties instead? Then the immutable ones wouldn't need to trigger a deep copy ever? In both of the cases you mention above, you are creating a read-only view so that you can either copy it or perform out of place operations on it IIUC.

Reply:
I too am worried about the potential for missteps when someone else now writes code related to manipulating buffers.

Perhaps we should do this in a followup cleanup, since it would be code changes without new functionality. By changing these properties to be accessors where one must explicitly request the access mode. That way the API for accessing is more uniform and it is clear at the call site how the buffer is being accessed. 

If I am honest, I have just read this PR, and already I can't keep straight in my head which property I want for "this is a read-only view, no need to copy" versus "this might be a third-party view that is written to, please copy eagerly" versus "this is an internal view that should be copied lazily".

Reply:
I want to limit the scope of this PR, but TBH I also don't want to merge it until we at least have a plan for how to handle this. The spilling code already makes buffer internals too hard to follow without some serious care. I think we need to put a hard cap on any changes that increase the complexity further without having a plan for how to fix it. Otherwise we will quickly find ourselves in a state where nobody will be able to debug this code without spending way too much time relearning each time (I still feel like I have to relearn the spilling internals every time I review a spilling PR, which I find quite concerning and I think we need to prioritize reducing that complexity and mitigating the fallout).